src/repo-service.cpp

```diff
@@ -138,3 +295,132 @@ void RepoService::openLocalFile(const QString& repo_id,
         helper->start();
     }
 }
+
+void RepoService::onGetRequestSuccess(const ServerRepo& repo)
+{
+    GetRepoRequest* req = qobject_cast<GetRepoRequest*>(sender());
+    if (!req)
+        return;
+
+    SyncedSubfolderhasRepoID subfolder_helper(repo.id);
+    std::vector<SyncedSubfolder>::iterator pos = std::find_if(synced_subfolders_.begin(), synced_subfolders_.end(), subfolder_helper);
+
+    if (pos != synced_subfolders_.end()) {
+        ServerRepo fixed_repo = repo;
+        fixed_repo.parent_path = pos->parent_path();
+        fixed_repo.parent_repo_id = pos->parent_repo_id();
+        server_repos_.push_back(fixed_repo);
+    }
+
+    delete req;
+    get_repo_reqs_.pop_front();
+
+    if (!get_repo_reqs_.empty())
+        get_repo_reqs_.front()->send();
+    else
+        emit refreshSuccess(server_repos_);
+}
+
+void RepoService::onGetRequestFailed(const ApiError& /*error*/)
+{
+    GetRepoRequest* req = qobject_cast<GetRepoRequest*>(sender());
+    if (!req)
+        return;
+
+    delete req;
+    get_repo_reqs_.pop_front();
+
+    if (!get_repo_reqs_.empty())
+        get_repo_reqs_.front()->send();
+    else
+        emit refreshSuccess(server_repos_);
+}
+
+void RepoService::startGetRequestFor(const QString &repo_id)
+{
+    bool was_empty = get_repo_reqs_.empty();
+    GetRepoRequest *req = new GetRepoRequest(seafApplet->accountManager()->currentAccount(), repo_id);
+    get_repo_reqs_.push_back(req);
+    connect(req, SIGNAL(success(const ServerRepo&)), this, SLOT(onGetRequestSuccess(const ServerRepo&)));
+    connect(req, SIGNAL(failed(const ApiError&)), this, SLOT(onGetRequestFailed(const ApiError&)));
+    if (was_empty)
+        req->send();
+}
+
+void RepoService::saveSyncedSubfolder(const ServerRepo& subfolder)
+{
+    const std::vector<Account>& accounts = seafApplet->accountManager()->accounts();
+    if (accounts.empty() || !subfolder.isSubfolder())
+        return;
+    const Account *account = &accounts.front();
+
+    // if we have it before
+    RepohasRepoID<ServerRepo> repo_helper(subfolder.id);
+    std::vector<ServerRepo>::iterator pos = std::find_if(server_repos_.begin(), server_repos_.end(), repo_helper);
+    if (pos == server_repos_.end()) {
+        server_repos_.push_back(subfolder);
+    } else {
+        pos->parent_path = subfolder.parent_path;
+        pos->parent_repo_id = subfolder.parent_repo_id;
+    }
+
+    if (!synced_subfolder_db_) {
+        qWarning("synced subfolder database is not available\n");
+        return;
+    }
+
+    QString sql = "REPLACE INTO SyncedSubfolder(repo_id, parent_repo_id, url, username, parent_path) VALUES ('%1', '%2', '%3', '%4', '%5')";
+    sql = sql.arg(subfolder.id)
+             .arg(subfolder.parent_repo_id)
+             .arg(account->serverUrl.toEncoded().data())
+             .arg(account->username)
+             .arg(subfolder.parent_path);
+    sqlite_query_exec (synced_subfolder_db_, sql.toUtf8().data());
```

Comment:
The path name may contain single quote `'` in it, so it's not safe to use `QString::arg()` to build the sql statement.  We could use `sqlite3_mprintf`.

https://www.sqlite.org/c3ref/mprintf.html